ampd/src/broadcaster/mod.rs

```diff
@@ -88,30 +104,120 @@ impl Default for Config {
+    async fn broadcast(&mut self, msgs: Vec<Any>) -> Result<TxResponse, Error>;
+    async fn estimate_fee(&mut self, msgs: Vec<Any>) -> Result<Fee, Error>;
 }
 
 #[derive(TypedBuilder)]
-pub struct BroadcastClient<T, S, Q> {
+pub struct UnvalidatedBasicBroadcaster<T, S, A, B>
+where
+    T: cosmos::BroadcastClient + Send,
+    S: Multisig + Send + Sync,
+    A: cosmos::AccountQueryClient + Send,
+    B: cosmos::BalanceQueryClient,
+{
     client: T,
     signer: S,
-    query_client: Q,
-    address: TMAddress,
+    auth_query_client: A,
+    bank_query_client: B,
+    address_prefix: String,
     #[builder(default, setter(skip))]
     acc_sequence: Option<u64>,
     pub_key: (String, PublicKey),
     config: Config,
 }
 
+impl<T, S, A, B> UnvalidatedBasicBroadcaster<T, S, A, B>
+where
+    T: cosmos::BroadcastClient + Send,
+    S: Multisig + Send + Sync,
+    A: cosmos::AccountQueryClient + Send,
+    B: cosmos::BalanceQueryClient,
+{
+    pub async fn validate_fee_denomination(mut self) -> Result<BasicBroadcaster<T, S, A>, Error> {
+        let denom: Denom = self.config.gas_price.denom.clone().into();
+        let address: TMAddress = self.derive_address()?;
+
+        ensure!(
+            self.balance(address.clone(), denom.clone())
+                .await?
+                .then(extract_non_zero_amount)
+                .is_some(),
+            Error::NoTokensOfFeeDenom { denom, address }
+        );
+
+        Ok(BasicBroadcaster {
+            client: self.client,
+            signer: self.signer,
+            auth_query_client: self.auth_query_client,
+            address: address.clone(),
+            acc_sequence: self.acc_sequence,
+            pub_key: self.pub_key,
+            config: self.config,
+        })
+    }
+
+    fn derive_address(&mut self) -> Result<TMAddress, Error> {
+        Ok(self
+            .pub_key
+            .1
+            .account_id(&self.address_prefix)
+            .change_context(Error::AddressEncoding)?
+            .into())
+    }
+
+    async fn balance(&mut self, address: TMAddress, denom: Denom) -> Result<Coin, Error> {
+        let coin = self
+            .bank_query_client
+            .balance(QueryBalanceRequest {
+                address: address.to_string(),
+                denom: denom.to_string(),
+            })
+            .await
+            .and_then(|response| {
+                response
+                    .balance
+                    .ok_or(Status::not_found("balance not found"))
+            })
+            .change_context(Error::QueryBalance { address, denom })?;
+
+        ResultCompatExt::change_context(
+            coin.try_into(),
+            Error::MalformedResponse {
+                query: "balance".to_string(),
+            },
+        )
+    }
+}
+
+fn extract_non_zero_amount(coin: Coin) -> Option<Amount> {
+    Some(coin.amount).filter(|amount| !amount.is_zero())
+}
+
+#[derive(Debug)]
+pub struct BasicBroadcaster<T, S, Q>
```

Comment:
Calling it `BasicBroadcaster` just so that it's name won't conflict with the trait?